Thinkful Bootcamp Course

Author: Ian Heaton

Email: iheaton@gmail.com

Mentor: Nemanja Radojkovic

Date: 2017/03/21


In [2]:
%matplotlib inline
import numpy as np
import pandas as pd
import scipy
import sklearn
import matplotlib.pyplot as plt
import seaborn as sb

sb.set_style('darkgrid')

# Using a Naive Bayes Classifier to look at Yelp Feedback  

## Sentiment Analysis

In [3]:
file = "/media/ianh/space/ThinkfulData/SentimentSentencesData/yelp_labelled.xlsx"
messages = pd.read_excel(open(file,'rb'), sheetname='yelp_labelled')

In [4]:
print("The number of missing data points for message : %d" % (sum(messages.message.isnull())))
print("The number of missing data points for result  : %d" % (sum(messages.result.isnull())))
print(messages.result.value_counts() / messages.shape[0])

The number of missing data points for message : 0
The number of missing data points for result  : 0
1    0.5
0    0.5
Name: result, dtype: float64


In [5]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.cross_validation import train_test_split

X_train, X_test, y_train, y_test = train_test_split(messages.message, messages.result, random_state=42)
vect = CountVectorizer()
train_matrix = vect.fit_transform(X_train)
test_matrix = vect.transform(X_test)

#Take a look at the resulting dataframe for training data
pd.DataFrame(train_matrix.toarray(), columns=vect.get_feature_names()).head()

/usr/local/lib/python3.5/dist-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


10  100  12  17  1979  20  2007  30  30s  35  ...   years  yellow  yet  \
0   0    0   0   0     0   0     0   0    0   0  ...       0       0    0   
1   0    0   0   0     0   0     0   0    0   0  ...       0       0    0   
2   0    0   0   0     0   0     0   0    0   0  ...       0       0    0   
3   0    0   0   0     0   0     0   0    0   0  ...       0       0    0   
4   0    0   0   0     0   0     0   0    0   0  ...       0       0    0   

   you  your  yourself  yucky  yum  yummy  zero  
0    0     0         0      0    0      0     0  
1    0     1         0      0    0      0     0  
2    0     0         0      0    0      0     0  
3    2     0         0      0    0      0     0  
4    0     0         0      0    0      0     0  

[5 rows x 1728 columns]

### Model Building With Naive Bayes

In [6]:
from sklearn.naive_bayes import BernoulliNB

#Create model
bnb = BernoulliNB()
# fit it to our training set
bnb.fit(train_matrix, y_train)
# make predictions on test data using test_dtm
predictions = bnb.predict(test_matrix)

### Accuracy of Model

In [7]:
# compare predictions to true results
from sklearn import metrics

print("Accuracy of the model is:  {:.2f}%\n".format(100 * metrics.accuracy_score(y_test, predictions)))
print("Confusion Matrix : \n", metrics.confusion_matrix(y_test, predictions))

Accuracy of the model is:  71.60%

Confusion Matrix : 
 [[80 48]
 [23 99]]


Our null accuracy was 50 % while our accuracy after using the classifier is at  72 %.
But our model has rather poor sensitivity and specificity with 48 messages being falsely classified as negative messages and 23 messages as being falsely classified as positive.  Perhaps Bayes classification is not the best model for the dataset.
